demos/fleet_provisioning/fleet_provisioning_with_csr/pkcs11_operations.c

```diff
@@ -0,0 +1,1162 @@
+
+        if( mbedKeyType == MBEDTLS_PK_RSA )
+        {
+            result = provisionPrivateRSAKey( session, label, &mbedPkContext );
+        }
+        else if( ( mbedKeyType == MBEDTLS_PK_ECDSA ) ||
+                 ( mbedKeyType == MBEDTLS_PK_ECKEY ) ||
+                 ( mbedKeyType == MBEDTLS_PK_ECKEY_DH ) )
+        {
+            result = provisionPrivateECKey( session, label, &mbedPkContext );
+        }
+        else
+        {
+            LogError( ( "Invalid private key type provided. RSA-2048 and EC P-256 keys "
+                        "are supported." ) );
+            result = CKR_ARGUMENTS_BAD;
+        }
+    }
+
+    mbedtls_pk_free( &mbedPkContext );
+
+    return result;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV provisionCertificate( CK_SESSION_HANDLE session,
+                                   uint8_t * certificate,
+                                   size_t certificateLength,
+                                   uint8_t * label )
+{
+    PKCS11_CertificateTemplate_t certificateTemplate;
+    CK_OBJECT_CLASS certificateClass = CKO_CERTIFICATE;
+    CK_CERTIFICATE_TYPE certificateType = CKC_X_509;
+    CK_FUNCTION_LIST_PTR functionList;
+    CK_RV result;
+    uint8_t * derObject = NULL;
+    int32_t conversion = 0;
+    size_t derLen = 0;
+    CK_BBOOL tokenStorage = CK_TRUE;
+    CK_BYTE subject[] = "TestSubject";
+    CK_OBJECT_HANDLE objectHandle = CK_INVALID_HANDLE;
+
+    /* Initialize the client certificate template. */
+    certificateTemplate.xObjectClass.type = CKA_CLASS;
+    certificateTemplate.xObjectClass.pValue = &certificateClass;
+    certificateTemplate.xObjectClass.ulValueLen = sizeof( certificateClass );
+    certificateTemplate.xSubject.type = CKA_SUBJECT;
+    certificateTemplate.xSubject.pValue = subject;
+    certificateTemplate.xSubject.ulValueLen = strlen( ( const char * ) subject );
+    certificateTemplate.xValue.type = CKA_VALUE;
+    certificateTemplate.xValue.pValue = ( CK_VOID_PTR ) certificate;
+    certificateTemplate.xValue.ulValueLen = ( CK_ULONG ) certificateLength;
+    certificateTemplate.xLabel.type = CKA_LABEL;
+    certificateTemplate.xLabel.pValue = ( CK_VOID_PTR ) label;
+    certificateTemplate.xLabel.ulValueLen = strlen( ( const char * ) label );
+    certificateTemplate.xCertificateType.type = CKA_CERTIFICATE_TYPE;
+    certificateTemplate.xCertificateType.pValue = &certificateType;
+    certificateTemplate.xCertificateType.ulValueLen = sizeof( CK_CERTIFICATE_TYPE );
+    certificateTemplate.xTokenObject.type = CKA_TOKEN;
+    certificateTemplate.xTokenObject.pValue = &tokenStorage;
+    certificateTemplate.xTokenObject.ulValueLen = sizeof( tokenStorage );
+
+    result = C_GetFunctionList( &functionList );
+
+    /* Test for a valid certificate: 0x2d is '-', as in ----- BEGIN CERTIFICATE. */
+    if( ( certificate == NULL ) || ( certificate[ 0 ] != 0x2d ) )
+    {
+        result = CKR_ATTRIBUTE_VALUE_INVALID;
+    }
```

Comment:
This seems like an insufficient certificate verification methodology. Should mbedTLS cert parsing be used instead?